src/contracts/exchange/NiftyswapExchange20.sol

```diff
@@ -0,0 +1,812 @@
+   */
+  function _removeLiquidity(
+    address _provider,
+    uint256[] memory _tokenIds,
+    uint256[] memory _poolTokenAmounts,
+    uint256[] memory _minCurrency,
+    uint256[] memory _minTokens,
+    uint256 _deadline)
+    internal nonReentrant()
+  {
+    // Input validation
+    require(_deadline > block.timestamp, "NiftyswapExchange#_removeLiquidity: DEADLINE_EXCEEDED");
+
+    // Initialize variables
+    uint256 nTokens = _tokenIds.length;                        // Number of Token IDs to deposit
+    uint256 totalCurrency = 0;                                 // Total amount of currency  to transfer
+    uint256[] memory tokenAmounts = new uint256[](nTokens);    // Amount of Tokens to transfer for each id
+    uint256[] memory currencyAmounts = new uint256[](nTokens); // Amount of currency to transfer for each id
+    uint256[] memory tokenReserves = new uint256[](nTokens);
+
+    // Get token reserves
+    tokenReserves = _getTokenReserves(_tokenIds);
+
+    // Assumes NIFTY liquidity tokens are already received by contract, but not
+    // the currency nor the Tokens Ids
+
+    // Remove liquidity for each Token ID in _tokenIds
+    for (uint256 i = 0; i < nTokens; i++) {
+      // Store current id and amount from argument arrays
+      uint256 id = _tokenIds[i];
+      uint256 amountPool = _poolTokenAmounts[i];
+      uint256 tokenReserve = tokenReserves[i];
+
+      // Load total liquidity pool token supply for Token _id
+      uint256 totalLiquidity = totalSupplies[id];
+      require(totalLiquidity > 0, "NiftyswapExchange#_removeLiquidity: NULL_TOTAL_LIQUIDITY");
+
+      // Load currency and Token reserve's supply of Token id
+      uint256 currencyReserve = currencyReserves[id];
+
+      // Calculate amount to withdraw for currency and Token _id
+      uint256 currencyAmount = amountPool.mul(currencyReserve) / totalLiquidity;
+      uint256 tokenAmount = amountPool.mul(tokenReserve) / totalLiquidity;
+
+      // Verify if amounts to withdraw respect minimums specified
+      require(currencyAmount >= _minCurrency[i], "NiftyswapExchange#_removeLiquidity: INSUFFICIENT_CURRENCY_AMOUNT");
+      require(tokenAmount >= _minTokens[i], "NiftyswapExchange#_removeLiquidity: INSUFFICIENT_TOKENS");
+
+      // Update total liquidity pool token supply of Token _id
+      totalSupplies[id] = totalLiquidity.sub(amountPool);
+
+      // Update currency reserve size for Token id
+      currencyReserves[id] = currencyReserve.sub(currencyAmount);
+
+      // Update totalCurrency and tokenAmounts
+      totalCurrency = totalCurrency.add(currencyAmount);
+      tokenAmounts[i] = tokenAmount;
+      currencyAmounts[i] = currencyAmount;
+    }
+
+    // Burn liquidity pool tokens for offchain supplies
+    _batchBurn(address(this), _tokenIds, _poolTokenAmounts);
+
+    // Transfer total currency and all Tokens ids
+    currency.transfer(_provider, totalCurrency);
```

Comment:
Should require that this returns true since not all ERC-20 transfer implementation will throw on error.